File: src/application/config/validation.ts
import { isGraphWorkflowConfig, type ProjectConfig } from "./types.js";

/** Post-parse invariant checks — not part of public façade exports. */
export function validateConfigReferences(config: ProjectConfig): void {
  for (const [agentId, agent] of Object.entries(config.agents)) {
    for (const [purpose, selection] of Object.entries(agent.models)) {
      if (
        selection !== "dynamic" &&
        !config.models.tiers[selection] &&
        selection.trim().length === 0
      ) {
        throw new Error(
          `Agent "${agentId}" has invalid model selection for ${purpose}: ${selection}`,
        );
      }
    }
  }

  for (const [workflowId, workflow] of Object.entries(config.workflows)) {
    if (isGraphWorkflowConfig(workflow)) {
      continue;
    }

    for (const agentId of workflow.agents) {
      if (!config.agents[agentId]) {
        throw new Error(`Workflow "${workflowId}" references unknown agent "${agentId}".`);
      }
    }

    if (workflow.qa && !config.agents[workflow.qa.agent]) {
      throw new Error(
        `Workflow "${workflowId}" references unknown QA agent "${workflow.qa.agent}".`,
      );
    }

    for (const tier of workflow.dispatch?.tiers ?? []) {
      for (const agentId of tier.agents) {
        if (!config.agents[agentId]) {
          throw new Error(
            `Workflow "${workflowId}" dispatch tier "${tier.name}" references unknown agent "${agentId}".`,
          );
        }
      }

      if (tier.qa && !workflow.qa) {
        throw new Error(
          `Workflow "${workflowId}" dispatch tier "${tier.name}" enables QA but no QA config is present.`,
        );
      }
    }
  }
}

export function validateMemoryBudget(config: ProjectConfig): void {
  if (config.memory.maxRamMb === "auto") {
    return;
  }
  const cap = config.memory.maxRamMb;
  const violations: string[] = [];
  for (const [tierId, tier] of Object.entries(config.models.tiers)) {
    if (tier.estimatedRamMb > cap) {
      violations.push(
        `models.tiers.${tierId}.estimatedRamMb (${tier.estimatedRamMb} MB) exceeds memory.maxRamMb (${cap} MB)`,
      );
    }
  }
  if (violations.length > 0) {
    throw new Error(violations.join("; "));
  }
}

File: crates/rlm-core/src/application/config/validation.rs
use std::io;

use serde_json::Value;

use crate::application::memory::validate_memory_budget as check_memory_budget;

pub(crate) fn validate_config_shape(config: &Value) -> io::Result<()> {
    let required = [
        "models",
        "memory",
        "runtime",
        "agents",
        "workflows",
        "interop",
        "hosts",
    ];
    let object = config.as_object().ok_or_else(|| {
        io::Error::new(
            io::ErrorKind::InvalidData,
            "expected config mapping at root",
        )
    })?;
    for key in required {
        if !object.contains_key(key) {
            return Err(io::Error::new(
                io::ErrorKind::InvalidData,
                format!("missing required config section: {key}"),
            ));
        }
    }
    Ok(())
}

pub(crate) fn validate_config_references(config: &Value) -> io::Result<()> {
    let tiers = config
        .pointer("/models/tiers")
        .and_then(Value::as_object)
        .cloned()
        .unwrap_or_default();
    if let Some(agents) = config.get("agents").and_then(Value::as_object) {
        for (agent_id, agent) in agents {
            if let Some(models) = agent.get("models").and_then(Value::as_object) {
                for (purpose, selection) in models {
                    let selection = selection.as_str().unwrap_or("");
                    if selection != "dynamic"
                        && !tiers.contains_key(selection)
                        && !selection.trim().is_empty()
                    {
                        return Err(io::Error::new(
                            io::ErrorKind::InvalidData,
                            format!(
                                "Agent \"{agent_id}\" has invalid model selection for {purpose}: {selection}"
                            ),
                        ));
                    }
                }
            }
        }
    }

    if let Some(workflows) = config.get("workflows").and_then(Value::as_object) {
        for (workflow_id, workflow) in workflows {
            if workflow.get("mode").and_then(Value::as_str) == Some("graph") {
                continue;
            }
            if let Some(agent_ids) = workflow.get("agents").and_then(Value::as_array) {
                for agent_id in agent_ids {
                    let agent_id = agent_id.as_str().unwrap_or("");
                    if !config
                        .pointer("/agents")
                        .and_then(Value::as_object)
                        .is_some_and(|agents| agents.contains_key(agent_id))
                    {
                        return Err(io::Error::new(
                            io::ErrorKind::InvalidData,
                            format!(
                                "Workflow \"{workflow_id}\" references unknown agent \"{agent_id}\"."
                            ),
                        ));
                    }
                }
            }
            if let Some(qa_agent) = workflow.pointer("/qa/agent").and_then(Value::as_str) {
                if !config
                    .pointer("/agents")
                    .and_then(Value::as_object)
                    .is_some_and(|agents| agents.contains_key(qa_agent))
                {
                    return Err(io::Error::new(
                        io::ErrorKind::InvalidData,
                        format!(
                            "Workflow \"{workflow_id}\" references unknown QA agent \"{qa_agent}\"."
                        ),
                    ));
                }
            }
        }
    }
    Ok(())
}

pub(crate) fn validate_memory_budget(config: &Value) -> io::Result<()> {
    check_memory_budget(config).map_err(|err| io::Error::new(io::ErrorKind::InvalidData, err))
}
